finish problem 1-3

hu-bin opened the pull request:
Change-Id: I5bc0e6d5986590c09e835a624483ad5f680c5a87

clone的时候用的bytedance的邮箱，而且没有先fork，所以有点混乱……

Commit:
finish problem 1-3

Change-Id: I5bc0e6d5986590c09e835a624483ad5f680c5a87

lifei commented on lesson1/department/service.py:
```diff
@@ -3,9 +3,68 @@
+    def is_user_in_department(self, employee_id, department_id) -> bool:
+        # raise NotImplementedError()
+
+        print("start function is_user_in_department ")
```
代码中不要有 print。有必要可以考虑打日志。

lifei commented on lesson1/department/service.py:
```diff
@@ -3,9 +3,68 @@
+        print("start function is_user_in_department ")
+        print("input employee_id:", employee_id, ", department_id:", department_id)
+
+        engine = create_engine(SQLALCHEMY_DATABASE_URI)
```
请使用 Flask-SQLAlchemy 来访问数据库。

lifei commented on lesson1/models.py:
```diff
@@ -29,3 +30,31 @@ class Permission(db.Model):
+    parent_path = db.Column(VARCHAR(128), nullable=False)
+
+
+class EmployeeDepartment(db.Model):
```
不要这个表。

lifei commented on lesson1/models.py:
```diff
@@ -29,3 +30,31 @@ class Permission(db.Model):
+    __name__ = 'department_tree'
+    # mysql_character_set = 'utf8mb4' -- 使用该语句可以建立字符集为utf8mb4的表
+    id = db.Column(BIGINT(unsigned=True), primary_key=True)
+    parent_path = db.Column(VARCHAR(128), nullable=False)
```
数据库表结构存储要满足范式

lifei commented on lesson1/models.py:
```diff
@@ -29,3 +30,31 @@ class Permission(db.Model):
+    department_id = db.Column(BIGINT(unsigned=True), nullable=False)
+
+
+class DepartmentPermission(db.Model):
```
一个表满足两个需求？不够明确。

Commit:
Fix problem：
1) replace print with log,
2) change mysql access method,
3) change database models.
Add feature: redis cache.

Change-Id: Ifea41a0fea8410cd479c4b769c64765456539125

lifei commented on lesson1/common/redis_helper.py:
```diff
@@ -0,0 +1,65 @@
+
+    # 尝试从redis读取
+    # r = redis.StrictRedis(host='172.18.0.2', port=6379, db=0)  # , decode_responses=True
+    r = redis.StrictRedis.from_url(REDIS_URL)
```
redis 不要这样用，这个变量应该是单例的。

lifei commented on lesson1/common/redis_helper.py:
```diff
@@ -0,0 +1,65 @@
+    # 如果取不到从数据库读取
+    if not parent_department:
+
+        parent_department = DepartmentClosure.get_parent_department(department_id)
```
> Closure

这个是什么意思？

lifei commented on lesson1/common/redis_helper.py:
```diff
@@ -0,0 +1,65 @@
+
+        # 添加到redis中
+        if parent_department is not None:
+            for i in parent_department:
```
> parent_department

这个变量看上去是单数，结果变成了复数。

lifei commented on lesson1/config/common.py:
```diff
@@ -4,3 +4,4 @@
 
 SQLALCHEMY_DATABASE_URI = 'mysql+pymysql://root:dev@mysql/lesson1?charset=utf8mb4'
 SQLALCHEMY_TRACK_MODIFICATIONS = False
+REDIS_URL = 'redis://{}:{}/0'.format('172.18.0.2', 6379)
```
> 172.18.0.2

使用 link 的别名。
